Add nonclobbering packages that only register if no other compressor has been registered with the same name

mostynb opened the pull request:
These new packages are suitable for other reusable libraries to import, if they do not want to accidentally clobber their user's previously registered compressors.

This is an alternative to #18 which does not change the behaviour for existing users- if you want the nonclobbering variant you must import that instead.

jmacd commented:
I think this is reasonable.

jmacd commented:
I slightly prefer always not clobbering, but not changing existing users is a safe way to move forward.

mostynb commented:
Added all the compressors and updated documentation. I haven't looked into adding tests for the new behaviour yet (maybe we can adapt the new tests in #18)- let's see what the opentelemtry-collector folks think first.

Filed https://github.com/open-telemetry/opentelemetry-collector/issues/7920 to see if opentelemetry-collector would accept this change on their side.

mostynb commented:
Landed this change with @jmacd's tests, and made a v1.2.0 release. Here's an opentelemetry-collector PR to use the new mode: https://github.com/open-telemetry/opentelemetry-collector/pull/7966

mostynb commented:
FYI, this new mode is now used in opentelemetry-collector: https://github.com/open-telemetry/opentelemetry-collector/pull/7966
